File: apollo_dump/_archived_20260605/locate_acquiresmd.js
'use strict';

// locate_acquiresmd.js — Frida 定位 DGraphicAcquireSMD 入口
// 原理:
//   1. Memory.scan(.text, "SSKF") → 定位函数体内魔数比较处
//   2. 从命中地址往回扫描 x86 函数序言 (push ebp; mov ebp, esp)
//   3. hook 候选入口, 验证是否被调用 + 捕获参数

var fsMod = Process.getModuleByName('FreeStyle.exe');
var base = fsMod.base;
var textSection = null;
fsMod.enumerateSections().forEach(function(s) {
    if (s.name === '.text') textSection = s;
});

var textStart = null;
var textSize = null;
var textEnd = null;
var FOUND_ADDR = null;
var HOOKED_ADDR = null;
var SSKF_COUNT = 0;

function boot() {
    if (!textSection) {
        send({t: 'FATAL', msg: '.text section not found'});
        return;
    }
    textStart = textSection.address;
    textSize = textSection.size;
    textEnd = textStart.add(textSize);
    setTimeout(run, 100);
}

function run() {
    send({t: 'INIT', msg: 'locate_acquiresmd.js — Scanning for DGraphicAcquireSMD entry...'});
    send({t: 'PHASE', msg: 'base=' + base + ' .text=' + textStart + ' size=' + textSize});

    Memory.scan(textStart, textSize, "53 53 4B 46", {
        onMatch: function(address, size) {
            SSKF_COUNT++;
            var offset = address.sub(base).toInt32();
            send({t: 'SSKF_HIT', offset: '0x' + offset.toString(16), addr: address.toString()});
            if (FOUND_ADDR === null) FOUND_ADDR = address;
        },
        onComplete: function() {
            send({t: 'PHASE', msg: 'SSKF scan done: ' + SSKF_COUNT + ' hits'});
            if (SSKF_COUNT === 0) {
                send({t: 'FATAL', msg: 'SSKF not found in .text!'});
                return;
            }
            setTimeout(findPrologue, 20);
        }
    });
}

function findPrologue() {
    send({t: 'PHASE', msg: 'Back-scanning for function prologue from SSKF (0x' + (FOUND_ADDR.sub(base).toInt32()).toString(16) + ')...'});

    var searchStart = FOUND_ADDR.sub(0x1000);
    if (searchStart.compare(textStart) < 0) searchStart = textStart;
    var searchLen = FOUND_ADDR.sub(searchStart).toInt32();

    send({t: 'PHASE', msg: 'Back-scan range: 0x' + (FOUND_ADDR.sub(base).toInt32()).toString(16) + ' - 0x1000 = 0x' + (searchStart.sub(base).toInt32()).toString(16) + ' (' + searchLen + ' bytes)'});

    var candidates = [];
    Memory.scan(searchStart, searchLen, "55", {
        onMatch: function(addr, size) {
            if (addr.compare(FOUND_ADDR) >= 0) return;
            var b2 = addr.add(1).readU8();
            var b3 = addr.add(2).readU8();
            var pattern = null;

            if (b2 === 0x8B && b3 === 0xEC) pattern = 'push ebp; mov ebp, esp';
            else if (b2 === 0x89 && b3 === 0xE5) pattern = 'push ebp; mov ebp, esp (alt)';
            else if (b2 === 0x83 && (b3 & 0xE0) === 0xE0) pattern = 'push ebp; sub esp, imm8';
            else if (b2 === 0x81) pattern = 'push ebp; sub esp, imm32';
            else if (b2 === 0xC8) pattern = 'enter XX, XX';

            if (pattern) {
                var offset = addr.sub(base).toInt32();
                var dist = FOUND_ADDR.sub(addr).toInt32();
                candidates.push({addr: addr, offset: offset, dist: dist, pattern: pattern});
                send({t: 'CANDIDATE', offset: '0x' + offset.toString(16), dist: dist, pattern: pattern});
            }
        },
        onComplete: function() {
            if (candidates.length > 0) {
                candidates.sort(function(a, b) { return a.dist - b.dist; });
                var best = candidates[0];
                HOOKED_ADDR = best.addr;
                send({t: 'SELECT', offset: '0x' + best.offset.toString(16), pattern: best.pattern, msg: 'Selected candidate @ FreeStyle+0x' + best.offset.toString(16) + ' [' + best.pattern + '] (' + best.dist + 'B before SSKF)'});
                setTimeout(hookCandidate, 20);
                return;
            }

            send({t: 'WARN', msg: 'No standard prologue. Scanning for FPO/hotpatch patterns...'});
            var fpoCandidates = [];
            Memory.scan(searchStart, searchLen, "81 EC", {
                onMatch: function(addr) {
                    if (addr.compare(FOUND_ADDR) >= 0) return;
                    var offset = addr.sub(base).toInt32();
                    var dist = FOUND_ADDR.sub(addr).toInt32();
                    fpoCandidates.push({addr: addr, offset: offset, dist: dist, pattern: 'sub esp, imm32 (FPO)'});
                },
                onComplete: function() {
                    if (fpoCandidates.length > 0) {
                        fpoCandidates.sort(function(a, b) { return a.dist - b.dist; });
                        var best = fpoCandidates[0];
                        HOOKED_ADDR = best.addr;
                        send({t: 'SELECT', offset: '0x' + best.offset.toString(16), pattern: best.pattern, msg: 'FPO candidate @ FreeStyle+0x' + best.offset.toString(16) + ' (' + best.dist + 'B before SSKF)'});
                        setTimeout(hookCandidate, 20);
                        return;
                    }

                    send({t: 'WARN', msg: 'No FPO either. Dumping hex around SSKF for manual analysis...'});
                    var dumpStart = FOUND_ADDR.sub(0x200);
                    if (dumpStart.compare(textStart) < 0) dumpStart = textStart;
                    var dumpBytes = [];
                    for (var i = 0; i < 0x200; i++) {
                        dumpBytes.push(dumpStart.add(i).readU8().toString(16).padStart(2, '0'));
                    }
                    send({t: 'HEXDUMP', offset: '0x' + dumpStart.sub(base).toInt32().toString(16), data: dumpBytes.join(' ')});
                    HOOKED_ADDR = FOUND_ADDR.sub(200);
                    send({t: 'SELECT', offset: '0x' + HOOKED_ADDR.sub(base).toInt32().toString(16), pattern: 'GUESS', msg: 'Guessing entry ~200B before SSKF @ FreeStyle+0x' + HOOKED_ADDR.sub(base).toInt32().toString(16)});
                    setTimeout(hookCandidate, 20);
                }
            });
        }
    });
}

function hookCandidate() {
    if (!HOOKED_ADDR) {
        send({t: 'FATAL', msg: 'No candidate address to hook'});
        return;
    }

    var hookedOffset = HOOKED_ADDR.sub(base).toInt32();
    send({t: 'PHASE', msg: 'Installing hook @ FreeStyle+0x' + hookedOffset.toString(16) + '...'});

    try {
        Interceptor.attach(HOOKED_ADDR, {
            onEnter: function(args) {
                var stack = Thread.backtrace(this.context, Backtracer.FUZZY).map(function(v) {
                    var mod = Process.findModuleByAddress(v);
                    if (mod) return mod.name + '+0x' + v.sub(mod.base).toInt32().toString(16);
                    return v.toString();
                }).join(' <- ');

                var argInfo = '';
                for (var i = 0; i < Math.min(args.length, 6); i++) {
                    try {
                        var ap = args[i];
                        if (ap.isNull()) { argInfo += ' arg' + i + '=NULL'; continue; }
                        var str = ap.readUtf16String();
                        if (str && str.length > 0 && str.length < 200) { argInfo += ' arg' + i + '="' + str + '"'; }
                        else { argInfo += ' arg' + i + '=' + ap.toString(); }
                    } catch(e) { argInfo += ' arg' + i + '=' + args[i].toString(); }
                }

                send({t: 'HOOK_HIT',
                    offset: '0x' + hookedOffset.toString(16),
                    args: argInfo,
                    tid: Process.getCurrentThreadId(),
                    stack: stack
                });
            },
            onLeave: function(retval) {
                send({t: 'HOOK_LEAVE', offset: '0x' + hookedOffset.toString(16), retval: retval.toString()});
            }
        });
        send({t: 'SUCCESS', msg: 'Hook live @ FreeStyle+0x' + hookedOffset.toString(16) + '. Waiting for calls...'});
    } catch(e) {
        send({t: 'ERROR', msg: 'Hook installation failed: ' + e});
    }
}

setTimeout(boot, 300);

File: apollo_dump/_archived_20260605/locate_acquiresmd.py
# locate_acquiresmd.py — Frida 定位 AcquireSMD 入口 + 日志导出
# 用法: py locate_acquiresmd.py
# 输出: locate_acquiresmd_YYYYMMDD_HHMMSS.txt

import frida
import sys
import os
import time
import subprocess
from datetime import datetime

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
JS_FILE = os.path.join(SCRIPT_DIR, 'locate_acquiresmd.js')
LOG_FILE = os.path.join(SCRIPT_DIR, 'locate_acquiresmd_' + datetime.now().strftime('%Y%m%d_%H%M%S') + '.txt')

log_handle = None

def log(msg):
    ts = datetime.now().strftime('%H:%M:%S.%f')[:12]
    line = f'[{ts}] {msg}'
    print(line)
    if log_handle:
        log_handle.write(line + '\n')
        log_handle.flush()

def find_pid():
    try:
        result = subprocess.run(
            ['tasklist', '/fi', 'IMAGENAME eq FreeStyle.exe', '/fo', 'csv', '/nh'],
            capture_output=True, text=True, timeout=10
        )
        for line in result.stdout.strip().split('\n'):
            if 'FreeStyle.exe' in line:
                return int(line.split(',')[1].strip('"'))
    except Exception as e:
        log(f'[!] tasklist failed: {e}')
    return None

def on_message(msg, data):
    if msg['type'] == 'send':
        p = msg['payload']
        t = p.get('t', '?')

        if t == 'INIT':
            log(f'[INIT] {p["msg"]}')
        elif t == 'INFO':
            log(f'[INFO] base={p["base"]} .text={p["textStart"]}-{p["textEnd"]} ({p["textSize"]} bytes)')
        elif t == 'PHASE':
            log(f'[PHASE] {p["msg"]}')
        elif t == 'SSKF_HIT':
            log(f'[SSKF] #{p["offset"]} @ {p["addr"]}')
        elif t == 'CANDIDATE':
            log(f'[CAND]  {p["msg"]}')
        elif t == 'SELECT':
            log(f'[SELECT] {p["msg"]}')
        elif t == 'HOOK_HIT':
            log(f'[HOOK] {p["msg"]}')
            log(f'[HOOK]   args[0]={p.get("args", "?")}')
            log(f'[HOOK]   stack={p.get("stack", "?")}')
        elif t == 'HOOK_LEAVE':
            log(f'[LEAVE] FreeStyle+{p["offset"]} retval={p["retval"]}')
        elif t == 'SUCCESS':
            log(f'[SUCCESS] {p["msg"]}')
        elif t == 'WARN':
            log(f'[WARN] {p["msg"]}')
        elif t == 'ERROR':
            log(f'[ERROR] {p["msg"]}')
        elif t == 'FATAL':
            log(f'[FATAL] {p["msg"]}')
        elif t == 'DONE':
            log(f'[DONE] {p["msg"]}')
        else:
            log(f'[{t}] {p.get("msg", str(p))}')
    elif msg['type'] == 'error':
        log(f'[JS ERROR] {msg.get("description", msg)}')

def main():
    global log_handle

    log_handle = open(LOG_FILE, 'w', encoding='utf-8')
    log(f'locate_acquiresmd.py started')
    log(f'Log file: {LOG_FILE}')
    log(f'')

    pid = find_pid()
    if pid is None:
        log('[!] FreeStyle.exe not found. Start game first.')
        log_handle.close()
        sys.exit(1)
    log(f'[+] FreeStyle.exe PID: {pid}')

    log(f'[*] Attaching via Frida...')
    try:
        session = frida.attach(pid)
    except Exception as e:
        log(f'[!] Attach failed: {e}')
        log_handle.close()
        sys.exit(1)

    with open(JS_FILE, 'r', encoding='utf-8') as f:
        js_code = f.read()

    script = session.create_script(js_code)
    script.on('message', on_message)
    script.load()
    log(f'[*] Script loaded. Waiting for SSKF scan results...')
    log(f'')

    log(f'[TIP] Go into a game scene (not lobby) so AcquireSMD is called.')
    log(f'[TIP] Press Ctrl+C to stop.')
    log(f'')

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        log(f'\n[EXIT] User interrupt')
    finally:
        try:
            session.detach()
        except:
            pass
        log(f'[DONE] Log saved to: {LOG_FILE}')
        log_handle.close()

if __name__ == '__main__':
    main()